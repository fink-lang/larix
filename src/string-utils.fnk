unindent_text = fn text, ind:
  [first, ...rest] = text.split rx/\n/g

  [...lines] = pipe rest:
    map line:
      line.slice ind

  [first, ...lines].join '\n'
