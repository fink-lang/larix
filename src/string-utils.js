import {curr_loc, next_loc} from '@fink/prattler';


export const unindent_text = (text, ind)=> (
  text
    .split(/\n/g)
    .map((line, idx)=> line.slice(idx > 0 ? ind : 0))
    .join('\n')
);


export const strip_block = (strings, ...parts)=> {
  const [, ...lines] = String.raw({raw: strings}, ...parts).split('\n');
  const ind = lines[0].search(/[^ ]/);
  const str = lines.map((lne)=> lne.slice(ind)).join('\n');
  return str;
};

