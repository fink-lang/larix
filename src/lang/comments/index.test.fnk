{describe, it, expect, to_match_snapshot} = import '@fink/jest'

{parse} = import '../../'


describe 'comments', fn:

  it 'parses line comments', fn:
    expect
      parse '
        # leading expression
        # multi line
        spam = ni

        # single line
        shrub = ni

        foo = fn:
          shrub
          # trailing func

        ni
      '
      to_match_snapshot


  it 'parses doc-comment', fn:
    expect
      parse '
        ---
        foobar spam ham
        ---
        spam = ni

        ---bar---
        spam
      '
      to_match_snapshot

