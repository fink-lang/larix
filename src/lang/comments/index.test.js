import {strip_block} from '../../string-utils';
import {parse_expr} from '../../';


describe('comments', ()=> {

  it('parses line comment', ()=> {
    expect(
      parse_expr(strip_block`
        # foobar
        spam = ni
      `)
    ).toMatchSnapshot();
  });


  it('parses doc-comment', ()=> {
    expect(
      parse_expr(strip_block`
        ---
        foobar
        ---
        spam = ni
      `)
    ).toMatchSnapshot();
  });
});
