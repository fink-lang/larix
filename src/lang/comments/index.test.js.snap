// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`comments parses doc-comment 1`] = `
assign = (4:0-4:9)
  ident (4:0-4:4) spam
  ident (4:7-4:9) ni
`;

exports[`comments parses line comment 1`] = `
assign = (2:0-2:9)
  ident (2:0-2:4) spam
  ident (2:7-2:9) ni
`;
